### PR TITLE
feat(cjs): allow users to wrap requireModule calls

### PR DESCRIPTION
**Pull request by RomanYarik**
Currently the users has no ability to know when a require call started or finished.

We want to allow the users the ability to hook into the evaluation process.

### CODE REVIEW

**Review by barak007 — CHANGES_REQUESTED** [review]
You must add more test cases to ensure the right behavior. Like deeper nesting. multiple times same module evaluation. Multiple requests that resolves to the same module. My debate is to maybe wrap the eval instead of the require. or add the cache to the wrap function that way you will be able to distinguish if the module is fresh without external registry. But maybe it's not needed. This is why more tests are needed.

File: packages/commonjs/src/managed-cjs-module-system.ts [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+
+      return resolvedFile;
+    },
+    dirname,
+    readFileSync,
+    wrapRequire: (requireCall, loadedModules) => {
+      const wrappedRequireCall = wrapRequire(requireCall, loadedModules);
+      return (modulePath) => {
+        if (modulePath) {
+          if (!moduleGraph.has(modulePath)) {
```
Comment (barak007) [inline packages/commonjs/src/managed-cjs-module-system.ts:64]: Combine ifs

File: packages/commonjs/src/managed-cjs-module-system.ts [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+  const requireFrom: ICommonJsModuleSystem['requireFrom'] = (contextPath, request) => {
+    const resolvedRequest = msResolveFrom(contextPath, request);
+    if (resolvedRequest === undefined) {
+      throw new Error(`Cannot resolve "${request}" in ${contextPath}`);
+    }
+
+    return requireModule(resolvedRequest);
+  };
+
+  const invalidateModule = (modulePath: string | false, visited = new Set<string>()) => {
```
Comment (barak007) [inline packages/commonjs/src/managed-cjs-module-system.ts:117]: Maybe we need pass filter here. Let's talk about it

Comment (barak007) [reply packages/commonjs/src/managed-cjs-module-system.ts:117]: Or even the strategy

File: packages/commonjs/src/managed-cjs-module-system.ts [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+};
+
+export const createManagedBaseCjsModuleSystem = ({
+  dirname,
+  readFileSync,
+  resolveFrom,
+  globals,
+  wrapRequire = (req) => (path) => req(path),
+}: IBaseModuleSystemOptions): IManagedCommonJsModuleSystem => {
+  const moduleGraph = new Map<
```
Comment (barak007) [inline packages/commonjs/src/managed-cjs-module-system.ts:32]: Let's think if we can expose the module graph

File: packages/commonjs/src/base-cjs-module-system.ts [hunk 4]
```diff
@@ -27,15 +29,24 @@ export interface IBaseModuleSystemOptions {
...
+  (path) =>
+    req(path);
+
 export function createBaseCjsModuleSystem(options: IBaseModuleSystemOptions): ICommonJsModuleSystem {
-  const { resolveFrom, dirname, readFileSync, globals = {} } = options;
+  const { resolveFrom, dirname, readFileSync, globals = {}, wrapRequire = defaultWrapRequire } = options;
 
   const loadedModules = new Map<string, IModule>();
 
+  const wrappedRequireModule = wrapRequire(requireModule, loadedModules);
```
Comment (barak007) [inline packages/commonjs/src/base-cjs-module-system.ts:46]: We don't want to wrap if we don't need it's just en extra function call and worth the less nice code

File: packages/commonjs/src/managed-cjs-module-system.ts [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+    string,
+    {
+      dependencies: string[];
+      importers: string[];
+    }
+  >();
+
+  const moduleSystem = createBaseCjsModuleSystem({
+    resolveFrom: (contextPath, request, origin) => {
```
Comment (barak007) [inline packages/commonjs/src/managed-cjs-module-system.ts:41]: We might want to expose resolveThrow

**Review by barak007 — COMMENTED** [review]
Generally looks good did the minimum modification to the core module system. And the wrapped one gives the most capabilities.

File: packages/commonjs/src/base-cjs-module-system.ts [hunk 6]
```diff
@@ -79,7 +85,18 @@ export function createBaseCjsModuleSystem(options: IBaseModuleSystemOptions): IC
...
-    const requireFromContext = (request: string) => requireFrom(contextPath, request, filePath);
+    const requireFromContext = (request: string) => {
+      const modulePath = resolveThrow(contextPath, request, filePath);
+      const moduleExports = requireModule(modulePath);
+      if (modulePath) {
+        const loadedModule = loadedModules.get(modulePath);
+        if (!loadedModule) {
+          throw new Error(`request ${request} failed to evaluate from ${filePath}`);
+        }
+        newModule.children.push(loadedModule);
```
Comment (barak007) [inline packages/commonjs/src/base-cjs-module-system.ts:96]: Chack if in node they appear twice

**Comment by RomanYarik** [comment]
re-implemented at #621

**Pull request closed without merging**